test/integration/integration_test.go

```diff
@@ -663,80 +668,120 @@ func awaitRolloutRestarts(t *testing.T, ctx context.Context,
 			}
 			return err
 		},
-		backoff.WithMaxRetries(backoff.NewConstantBackOff(time.Second*1), 60),
+		backoff.WithMaxRetries(backoff.NewConstantBackOff(time.Second*1), 360),
 	))
 }
 
+type status struct {
+	Replicas      *int32
+	ReadyReplicas int32
+}
+
+// assertRolloutRestarts asserts status and object state of each RolloutRestartTarget based on minGeneration.
+// If minGeneration==1, minimally assert each RolloutRestartTarget reached a healthy state after created
+// RolloutRestartTarget like argo.Rollout needs to reach a healthy state before we can effectively test
+// the rollout restart via a reconciler.Reconcile() - when argo.Rollout is making progress towards a healthy state
+// after created, any immediately triggered restart won't take effect
+// If minGeneration==2, assert each RolloutRestartTarget restarted, reached a healthy state, and has minGeneration
+// and appropriate restartAt.
 func assertRolloutRestarts(
 	t *testing.T, ctx context.Context, client ctrlclient.Client, obj ctrlclient.Object,
 	targets []secretsv1beta1.RolloutRestartTarget, minGeneration int64,
 ) error {
 	t.Helper()
 
-	type status struct {
-		Replicas      *int32
-		ReadyReplicas int32
-	}
 	var errs error
 	var s *status
 
 	// see secretsv1beta1.RolloutRestartTarget for supported target resources.
 	timeNow := time.Now().UTC()
+	var restartAt time.Time
 	for _, target := range targets {
 		var tObj ctrlclient.Object
 		tObjKey := ctrlclient.ObjectKey{
 			Namespace: obj.GetNamespace(),
 			Name:      target.Name,
 		}
 		var annotations map[string]string
+		expectedAnnotation := helpers.AnnotationRestartedAt
 		switch target.Kind {
 		case "Deployment":
-			var o appsv1.Deployment
-			if assert.NoError(t, client.Get(ctx, tObjKey, &o)) {
-				annotations = o.Spec.Template.Annotations
-				tObj = &o
+			if minGeneration > 1 {
+				var o appsv1.Deployment
+				if assert.NoError(t, client.Get(ctx, tObjKey, &o)) {
+					annotations = o.Spec.Template.Annotations
+					tObj = &o
+				}
+				s = &status{o.Spec.Replicas, o.Status.ReadyReplicas}
 			}
```

Comment:
Am I reading it right that `assertRolloutRestarts() `now skips checking Deployments, StatefulSets, or ReplicaSets at all unless `minGeneration > 1`? From the [calls in the dynamic secrets tests](https://github.com/hashicorp/vault-secrets-operator/blob/d4a768e8abde9e0d687709c78ed5bce78c027230/test/integration/vaultdynamicsecret_integration_test.go#L319-L323), it looks like `awaitRolloutrestarts()` is expected to verify that the targets are deployed and ready if `minGeneration=1`, but I'm not seeing that here.

https://github.com/hashicorp/vault-secrets-operator/blob/d4a768e8abde9e0d687709c78ed5bce78c027230/test/integration/vaultdynamicsecret_integration_test.go#L319-L323

Reply:
After creation, the generation is 1 and there is no restartAt due to no secret syncs happening yet, so my thought was we could skip checking those. My idea mainly was to reuse `awaitRolloutrestarts()` and `assertRolloutRestarts()` to ensure the argo.Rollout resource reaches a healthy state. This led to using `minGeneration=1` to control `assertRolloutRestarts()`'s minimal checking (no generation and restartAt checking).

Looking back I think this pattern is odd. I've removed `if minGeneration > 1 {` condition. We will check the generation but not restartAt in the case `minGeneration=1`.

Reply:
Resolving this conversation since I've reverted assertRolloutRestarts()'s behavior based on `minGeneration`.